opendevin/runtime/plugins/agent_skills/aider/linter.py

```diff
@@ -0,0 +1,223 @@
+import os
+import subprocess
+import sys
+import traceback
+import warnings
+from dataclasses import dataclass
+from pathlib import Path
+
+from grep_ast import TreeContext, filename_to_lang
+from tree_sitter_languages import get_parser  # noqa: E402
+
+# tree_sitter is throwing a FutureWarning
+warnings.simplefilter('ignore', category=FutureWarning)
+
+
+@dataclass
+class LintResult:
+    text: str
+    lines: list
+
+
+class Linter:
+    def __init__(self, encoding='utf-8', root=None):
+        self.encoding = encoding
+        self.root = root
+
+        self.languages = dict(
+            python=self.py_lint,
+        )
+        self.all_lint_cmd = None
+
+    def set_linter(self, lang, cmd):
+        if lang:
+            self.languages[lang] = cmd
+            return
+
+        self.all_lint_cmd = cmd
+
+    def get_rel_fname(self, fname):
+        if self.root:
+            return os.path.relpath(fname, self.root)
+        else:
+            return fname
+
+    def run_cmd(self, cmd, rel_fname, code):
+        cmd += ' ' + rel_fname
+        cmd = cmd.split()
+
+        process = subprocess.Popen(
+            cmd, cwd=self.root, stdout=subprocess.PIPE, stderr=subprocess.STDOUT
+        )
+        stdout, _ = process.communicate()
+        errors = stdout.decode().strip()
+        self.returncode = process.returncode
+        if self.returncode == 0:
+            return  # zero exit status
+
+        cmd = ' '.join(cmd)
+        res = ''
+        res += errors
+        line_num = extract_error_line_from(res)
+        return LintResult(text=res, lines=[line_num])
+
+    def get_abs_fname(self, fname):
+        if os.path.isabs(fname):
+            return fname
+        elif os.path.isfile(fname):
+            rel_fname = self.get_rel_fname(fname)
+            return os.path.abspath(rel_fname)
+        else:  # if a temp file
+            return self.get_rel_fname(fname)
+
+    def lint(self, fname, cmd=None) -> LintResult | None:
+        code = Path(fname).read_text(self.encoding)
+        absolute_fname = self.get_abs_fname(fname)
+        if cmd:
+            cmd = cmd.strip()
+        if not cmd:
+            lang = filename_to_lang(fname)
+            if not lang:
+                return None
+            if self.all_lint_cmd:
+                cmd = self.all_lint_cmd
+            else:
+                cmd = self.languages.get(lang)
+        if callable(cmd):
+            linkres = cmd(fname, absolute_fname, code)
+        elif cmd:
+            linkres = self.run_cmd(cmd, absolute_fname, code)
+        else:
+            linkres = basic_lint(absolute_fname, code)
+        return linkres
+
+    def flake_lint(self, rel_fname, code):
+        fatal = 'E9,F821,F823,F831,F406,F407,F701,F702,F704,F706'
```

Comment:
Is the performance degradation because we make the link much stricter and the agent failed to fix some of them and stuck in loops?

I think the original ones are just `F821,F822,F831,E112,E113,E999,E902`

Reply:
Yeah, might be... We might actually want to have only syntax error checks by default if we want to achieve the best pass rate? 

Then we could possibly make the syntax checks configurable so people could specify the ones they want to use for their specific project